Airspeed Velocity (asv) performance benchmarks setup: part-1

tanujkhattar opened the pull request:
This commit is the first step towards adding a performance benchmarking suit using asv. For an example of how the final benchmark page would look like, see https://tanujkhattar.github.io/Cirq/

This commit does the following:

* Migrates existing performance benchmarks to asv (rewrite in a format that asv understands)
* Adds an asv config file s.t. we can run the newly written performance benchmarks using asv

## Usage
### Step-1
Install asv and virtualenv
```bash
pip install asv        
pip install virtualenv
```
### Step-2
Add examples/ directory to an environment variable so that `benchmarks/bench_examples.py` can import the examples/ module. 
```bash
export ASV_PYTHONPATH="$(git rev-parse --show-toplevel)/examples"
```
**Note:** This hack is ugly and let's discuss whether we want to keep this or find an alternative. 
It is currently needed because:
* We are currently storing all benchmark tests in a separate benchmarks/* directory
* ASV tries to discover the imported modules during setup and is unable to find examples/* module because, unlike cirq/, it's not installed when we call the setup.py

An alternative could be to organise benchmark tests similar to pytests -- i.e. have a bench_<file_name>.py at the same path where <filename>.py is situated. If we do this, the `benchmark_dir` to asv can be specified as the root directory and it'll recursively search for all bench_* tests. (I haven't tested this but IIUC, this should work). 
We could also make it easier to run the asv tests by adding the necessary commands to `./check/performance` and then the users wouldn't have to worry about the hack if they use the script to run asv tests. 

### Step-3
Run the tests from Cirq root directory 
```
asv run
```

## TODOs
* Integrate this with GCP and setup cron job and infra to host the results webpage. -- this is blocked on merging this PR so that the tests and config file are present on the master branch. 
* Remove pytest-benchmarks setup and modify `./check/performance`.

vtomole commented on dev_tools/conf/.coveragerc:
```diff
@@ -3,3 +3,6 @@
 include = ./*
+
+omit = 
+    benchmarks/*
```
Nit: Add newline

tanujkhattar replied:
done

vtomole commented on asv.conf.json:
```diff
@@ -0,0 +1,160 @@
+{
```
Why is everything red?

tanujkhattar replied:
Because json doesn't support comments by default.

dstrain115 commented on asv.conf.json:
```diff
@@ -0,0 +1,160 @@
+    // "repo_subdir": "",
+
+    // Customizable commands for building, installing, and
+    // uninstalling the project. See asv.conf.json documentation.
```
Why does it say "See asv.conf.json documentation."  This file is asv.conf.json.

tanujkhattar replied:
This file template (and the comments) were auto-generated by asv -- but I think it's refering to the documentation online i.e. https://asv.readthedocs.io/en/stable/asv.conf.json.html#build-command-install-command-uninstall-command

tanujkhattar replied:
modified the comment to include the link.

dstrain115 commented on asv.conf.json:
```diff
@@ -0,0 +1,160 @@
+    // the base URL to show a commit for the project.
+    "show_commit_url": "https://github.com/quantumlib/Cirq/commit/",
+
+    // The Pythons you'd like to test against.  If not provided, defaults
```
nit: The python versions to test against.

tanujkhattar replied:
done.

dstrain115 commented on benchmarks/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020 The Cirq Developers
```
It's 2021 now.

tanujkhattar replied:
fixed.

dstrain115 commented on benchmarks/bench_examples.py:
```diff
@@ -0,0 +1,61 @@
+import examples.superdense_coding
+
+
+class ExamplesTest:
```
We should have comments here or at the top of the file to explain what this file is doing.  Same with bench_linalg_decompositions.py

tanujkhattar replied:
Hmm, so I could write a trivial comment like "Performance benchmarks for examples/*", but I think the situation is similar to how for every <filename>.py, we write unit tests in <filename>_test.py and don't add explicit comments unless needed otherwise. 

Maybe a benchmarks/README.md, explaining the setup and how to write / run tests, would be more appropriate? (I plan to add this in part-2 of the commit)

dstrain115 replied:
How about explaining more of the context?  There is no explanation of how to use this benchmark or its connection to "asv" or anything.  Pointing to a package-level README.md is okay.

tanujkhattar replied:
added a package-level README.md explaining the setup and usage of benchmarks.

tanujkhattar replied:
added docstring.

balopat commented on asv.conf.json:
```diff
@@ -0,0 +1,160 @@
+
+    // The Python versions you'd like to test against.  If not provided, defaults
+    // to the current version of Python used to run `asv`.
+    "pythons": ["3.6"],
```
I would do benchmarks on the latest python that we support (currently 3.8, soon 3.9).

balopat replied:
Optionally maybe we could do all our supported python versions? That would give an idea if there are spots in the codebase that have big differences between python version. (we can just create an issue for this)

tanujkhattar replied:
Add support for all versions (3.6, 3.7, 3.8) -- though these versions would have to be installed on the machine to be able to run tests for all of them.

balopat commented on asv.conf.json:
```diff
@@ -0,0 +1,160 @@
+{
+    // The version of the config file format.  Do not change, unless
```
In general, I would remove all the instructive comments from the JSON here and link to the original instructions JSON for reference. Especially the big unused, commented out sample JSON sections. 

I'm not against comments (even though they look terrible on Github :)) - if we want to add Cirq specific comments we are welcome to, which I think will be pretty minimal at this point.

tanujkhattar replied:
I personally would have preferred to keep the comments in the file so that the next time we have to edit it, the comments come in handy. 
Anyways, removed them.

balopat reviewed (changes requested):
Looks good:

- let's cleanup the JSON comments to keep only "interesting" / Cirq specific comments and link to the reference instead 
- We need more documentation around "How to do benchmarking" in Cirq - it could be a follow-up PR (this case please open an issue for it so we won't forget), especially that we don't have any docs for the current performance tests :) 
- Can we remove the check/performance script and the pytest-benchmark perf tests + the pytest benchmark plugin? I don't think we need it anymore if we introduce this suite.

tanujkhattar commented:
1. Cleaned up the json
2. Added a README.md in benchmarks/ explaining the setup and usage of asv for benchmark tests.
3. Removed check/performance script and pytest-benchmark perf tests + plugin.

I had initially planned to do 2 & 3 in a separate PR so that we can first get the asv stuff in and unblock the gcp setup. Included both as per the suggestion.

balopat reviewed (approved):
LGTM

dstrain115 commented on benchmarks/README.md:
```diff
@@ -0,0 +1,22 @@
+Please refer [Running Benchmarks guide by ASV](https://asv.readthedocs.io/en/stable/using.html#running-benchmarks) for more information. 
+
+## Results Database
+By default, the benchmarks are run on every commit to master and the results are stored in a GCP server and hosted on (TODO: add link after GCP setup). 
```
Let's delete the last part and link the TODO to a github issue.

tanujkhattar replied:
done

dstrain115 commented on benchmarks/bench_linalg_decompositions.py:
```diff
@@ -0,0 +1,40 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Worth a short docstring to say why KAK decomposition is the important thing to benchmark?

tanujkhattar replied:
Only KAK decomposition is currently benchmarked because it historically caused a problem of being too slow ([#1124](https://github.com/quantumlib/Cirq/issues/1124)) and this PR only ports the existing benchmarks to ASV. 

In future, we ideally should add benchmarks for other decompositions as well? I don't have a good reason as to why only KAK decomposition is important.

tanujkhattar replied:
Added an issue to track this: https://github.com/quantumlib/Cirq/issues/3840

tanujkhattar replied:
added a docstring.
